sys/include/json.h

```diff
@@ -0,0 +1,859 @@
+    JSON_WRITE_STATE_IN_STRING, /**< json_write_string() was last called */
+    JSON_WRITE_STATE_IN_OBJECT, /**< json_write_object_open() was last called */
+    JSON_WRITE_STATE_IN_ARRAY,  /**< json_write_array_open() was last called */
+    JSON_WRITE_STATE_IN_KEY,    /**< json_write_object_key() was last called */
+    JSON_WRITE_STATE_DONE,      /**< json_write_finish() was called  */
+} json_write_state_t;
+
+/**
+ * @brief   The cookie with all the information needed to serialize JSON data.
+ *
+ * @see @ref container_of
+ * @see json_write_init()
+ */
+struct json_write_cookie {
+    json_write_t write;       /**< The write function passed to json_write_init(). */
+    json_write_state_t state; /**< The internal state of the stream. */
+};
+
+/**
+ * @brief   Initializer for JSON serialization streams.
+ *
+ * @param[out] cookie           The cookie to subsequently use.
+ * @param[in] write             The write function to call back.
+ * @see json_write_finish()
+ */
+static inline void json_write_init(json_write_cookie_t *cookie, json_write_t write)
+{
+    cookie->write = write;
+    cookie->state = JSON_WRITE_STATE_SOMEWHERE;
+}
+
+/**
+ * @brief End a JSON serialization stream.
+ *
+ * @note An empty stream without any data is invalid.
+ *
+ * @see json_write_init()
+ *
+ * @param[in] cookie            The cookie, initialized with json_write_init().
+ *
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ */
+json_result_t json_write_finish(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Open an object to serialize.
+ *
+ * Inside the object call json_write_object_key() to write a key, then invoke
+ * any other write function to write the value associated to this key. Repeat
+ * until you are done, then call json_write_object_close(). An empty object
+ * is valid, too.
+ *
+ * @see json_write_object_close()
+ * @see json_write_object_key()
+ *
+ * @param[in] cookie            The cookie, initialized with json_write_init().
+ *
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_object_open(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Closes an object.
+ *
+ * @see json_write_object_open()
+ * @see json_write_object_key()
+ *
+ * @param[in] cookie            The cookie, initialized with json_write_init().
+ *
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_object_close(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Write a key inside an object.
+ *
+ * Invoke this function multiple times directly to concatenate this substring.
+ * You could write out a string `"test"` character by character if you want to.
```

Comment:
If the first sentence would be clearer this (somewhat confusing) example wouldn't be necessary.